homeassistant/components/light/__init__.py

```diff
@@ -287,9 +294,12 @@ async def async_handle_light_on_service(light, call):
 
             preprocess_turn_on_alternatives(hass, params)
 
-        if ATTR_PROFILE not in params:
+        if not params or not light.is_on:
```

Comment:
While I get why... this can cause unexpected behavior with lights that have an assumed state or, are transitioning/turn-on/off in between polling. This would affect e.g., Philips Hue or Zigbee lights that require polling.

Reply:
Can you elaborate on how this will affect them? I've spent some time looking into it but have been unable to determine the issue.

Reply:
This relies on the state in Home Assistant, which doesn't have to be accurate (as it depends on e.g., assumed state or the polling interval).

Reply:
Thank you for the clarification. Unfortunately, I'm not aware of a way we could get around this issue. I also believe this issue already presently exists, and has always existed with light profiles. I personally think we have to assume the Home Assistant state is accurate. If there is a better way, please let me know.

Reply:
> Unfortunately, I'm not aware of a way we could get around this issue.

There isn't away around it.

> I also believe this issue already presently exists

No, it didn't. You are no introducing one in this PR. As the light profile now relies on the state, previously, it just applied it (which is the whole reason you are doing this).

> I personally think we have to assume the Home Assistant state is accurate

We should not assume, especially not on this level. If an integration decides to do that locally, I would accept and agree, but this applies to ALL.

So, let's say we toggle a light twice, within the polling time of light integration. For the example, let's say the polling time is 10 seconds, the transition time is 5 seconds by default. and we are only using service calls with no parameters and the light is turned on now.

- Turn off the light: Turns off the light.... Transition will be 5 seconds + profile.
- Wait 3 seconds
- Turn on the light: Turns on the light... Transition is applied (as that doesn't have "must be on" condition), but no profile is applied either.

So what went wrong? Well.. During the transition of the first turn off (5 seconds) the light will be actually ON. (transition = ON, you can see it physically as well). However, the polling takes between 5 and 15 seconds (transition + polling) to realize that it is actually off.

However, we turn it on again after 3 seconds...  With his change it thinks is on already (but it is still on?), thus no profile is applied.

This can be easily be triggered with e.g., Hue blubs. Trying to blink or to some kind of alerting using the toggle service.

And this is why we generally send a complete set of attributes via a turn-on call.

Reply:
Actually just using the toggle service on a polling light (which we have a lot of), will just break this logic.

Reply:
Perhaps I'm misunderstanding, the toggle service already relies on the light.is_on state. https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py#L375-L381

For clarity, the default transition time still needs to be explicitly set by the user for a light entity, its not an across the board thing unless the user specifies an all lights group.

Reply:
Toggle will work in this case indeed. The bigger story still holds though.

Reply:
To be honest, I'm out of my depth in regards to how a polling light (as I do not have any) will react to calling the functions as you described.

However, logically, I believe if the `light.toggle` service works properly for a polling light with a transition time, then this code should work as well, as they both rely on the same `light.is_on` attribute.

Reply:
> (I'm not saying this code is wrong... it more that I'm trying to figure out the implication it might have and if that is wished-for).

I'm glad that you are criticizing this PR. It shows the effort and desire for high quality that you and other maintainers put into this open source project. So thank you :).